File: android/src/main/java/com/lightningkite/khrysalis/net/RxHttpAssist.actual.kt
package com.lightningkite.khrysalis.net

import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.module.kotlin.jacksonTypeRef
import com.fasterxml.jackson.module.kotlin.readValue
import com.lightningkite.khrysalis.*
import com.lightningkite.khrysalis.bytes.Data
import io.reactivex.Observable
import io.reactivex.Single
import java.lang.reflect.ParameterizedType

fun Single<@SwiftExactly("Element") HttpResponse>.unsuccessfulAsError(): Single<HttpResponse> {
    return this.map { it ->
        if(it.isSuccessful){
            return@map it
        } else {
            throw HttpResponseException(it)
        }
    }
}


fun Single<@SwiftExactly("Element") HttpResponse>.discard(): Single<Unit> {
    return this.flatMap { it.discard() }
}
inline fun <reified T> Single<@SwiftExactly("Element") HttpResponse>.readJson(): Single<T> {
    val type = jacksonTypeRef<T>()
    return this.flatMap { it ->
        if(it.isSuccessful){
            it.readJson<T>(type)
        } else {
            Single.error<T>(HttpResponseException(it)) as Single<T>
        }
    }
}
inline fun <reified T> Single<@SwiftExactly("Element") HttpResponse>.readJsonDebug(): Single<T> {
    val type = jacksonTypeRef<T>()
    return this.flatMap { it ->
        if(it.isSuccessful){
            it.readJsonDebug<T>(type)
        } else {
            Single.error<T>(HttpResponseException(it)) as Single<T>
        }
    }
}
fun Single<@SwiftExactly("Element") HttpResponse>.readText(): Single<String> {
    return this.flatMap { it ->
        if(it.isSuccessful){
            it.readText()
        } else {
            Single.error<String>(HttpResponseException(it))
        }
    }
}
fun Single<@SwiftExactly("Element") HttpResponse>.readData(): Single<Data> {
    return this.flatMap { it ->
        if(it.isSuccessful){
            it.readData()
        } else {
            Single.error<Data>(HttpResponseException(it))
        }
    }
}

File: ios/Khrysalis/src/net/RxHttpAssist.actual.swift
//Package: com.lightningkite.khrysalis.net
//Converted using Khrysalis2

import Foundation
import RxSwift
import RxRelay



//--- Single<@swiftExactly("Element")HttpResponse>.unsuccessfulAsError()
extension Single where Element == HttpResponse, Trait == SingleTrait {
    public func unsuccessfulAsError() -> Single<HttpResponse> {
        return self.map{ (it) in
            if it.isSuccessful {
                return it
            } else {
                throw HttpResponseException(it)
            }
        }
    }
}
 
 
 

extension PrimitiveSequence where Element == HttpResponse, Trait == SingleTrait {
    //--- Single<@swiftExactly("Element")HttpResponse>.discard()
    public func discard() -> Single<Void> {
        return self.flatMap { (it) in
            if it.isSuccessful {
                return it.discard()
            } else {
                return Single.error(HttpResponseException(it))
            }
        }
    }
    //--- Single<@swiftExactly("Element")HttpResponse>.readJson()
    public func readJson<T: Codable>() -> Single<T> {
        return self.flatMap { (it) in
            if it.isSuccessful {
                return it.readJson()
            } else {
                return Single.error(HttpResponseException(it))
            }
        }
    }
    //--- Single<@swiftExactly("Element")HttpResponse>.readJsonDebug()
    public func readJsonDebug<T: Codable>() -> Single<T> {
        return self.flatMap { (it) in
            if it.isSuccessful {
                return it.readJsonDebug()
            } else {
                return Single.error(HttpResponseException(it))
            }
        }
    }
    public func readJson<T: Codable>(_ type: T.Type) -> Single<T> {
        return self.flatMap { (it) in
            if it.isSuccessful {
                return it.readJson()
            } else {
                return Single.error(HttpResponseException(it))
            }
        }
    }
    //--- Single<@swiftExactly("Element")HttpResponse>.readText()
    public func readText() -> Single<String> {
        return self.flatMap { (it) in
            if it.isSuccessful {
                return Single.just(String(data: it.data, encoding: .utf8)!)
            } else {
                return Single.error(HttpResponseException(it))
            }
        }
    }
    //--- Single<@swiftExactly("Element")HttpResponse>.readData()
    public func readData() -> Single<Data> {
        return self.flatMap { (it) in
            if it.isSuccessful {
                return Single.just(it.data)
            } else {
                return Single.error(HttpResponseException(it))
            }
        }
    }
}
